rfd/0037-desktop-access-protocol.md

```diff
@@ -0,0 +1,191 @@
+---
+authors: Andrew Lytvynov (andrew@goteleport.com)
+state: draft
+---
+
+# RFD 37 - Desktop Access - Desktop protocol
+
+## What
+
+RFD 33 defines the high-level goals and architecture for Teleport Desktop
+Access.
+
+This RFD specifies the Desktop protocol - wire protocol used between the
+OS-specific Teleport desktop service (like `windows_desktop_service`) and the
+web client.  OS-specific Teleport desktop services are responsible for
+translating native protocols or APIs (like RDP or X11) into the protocol
+described here.
+
+## Details
+
+### Goals
+
+This custom protocol is created with several goals in mind:
+- performance - the messages should be compact and fast to encode/decode
+- portability - should map easily to standard protocols like RDP
+- decoding simplicity - the parsing code should be simple enough for auditing,
+  especially when written in dynamic languages like JavaScript
+- extendability - new capabilities can be added in the future
+
+### Overview
+
+The protocol consists of if discrete messages, sent between a client and a
+server. These messages are passed over a secure, authenticated and reliable
+transport layer, like TLS or a websocket. The protocol leaves all security
+concerns (authentication, integrity, etc) to the transport layer.
+
+Messages are bi-directional and asynchronous - client and server can send any
+message at any time to the other end. The expected sequence of messages for a
+typical desktop connection is described [below](#message-sequence).
+
+### Message sequence
+
+Typical sequence of messages in a desktop session:
+
+```
++--------+                                +--------+
```

Comment:
I had success using improbable engineering grpc over websockets transport:

https://github.com/improbable-eng/grpc-web

It supports streaming, typescript and could save you some time

Reply:
`grpc-web` right now does not support client-side streaming: https://github.com/improbable-eng/grpc-web#client-side-streaming
so for sending user input (mouse movement/clicks, keyboard) we'd need to use unary RPCs
I'm worried that this will add too much overhead and make the input even more sluggish than it is over a websocket.

I think the protocol described here is going to be easy to implement.
We can always swap it for gRPC later on, for maintainability - it'll simply mean adding a new proxy endpoint for the gRPC version of the protocol.